### PR TITLE
initial stuff

### PR DESCRIPTION
**Pull request by maciejATcosphere**

### CODE REVIEW

File: date_a_scientist/__init__.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class DateAScientist:
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        llm_openai_api_token: str | None = None,
+        llm_openai_model: str = "gpt-4o",
+        field_descriptions: dict[str, str] | None = None,
```
Comment (androj07) [inline date_a_scientist/__init__.py]: Maybe it would be worth mentioning this param as well in README

File: date_a_scientist/__init__.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    ) -> None:
+        self._df = df
+        self._field_descriptions = field_descriptions
+        self._llm_openai_api_token = llm_openai_api_token
+        self._llm_openai_model = llm_openai_model
+
+    def chat(self, q: str) -> Any:
+
+        result = self._agent.chat(self._query(q))
+        pattern = r"/[^\s]+"
```
Comment (androj07) [inline date_a_scientist/__init__.py]: What is this pattern for ? Is it for finding file paths in the result? Maybe we could consider more secific name?

File: date_a_scientist/__init__.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        content = self._add_plt_close(content)
+
+        return content
+
+    def _add_plt_close(self, text: str) -> str:
+        if "plt.savefig" in text:
+            lines = text.split("\n")
+            for i, line in enumerate(lines):
+                if line.startswith("plt.savefig"):
+                    lines.insert(i + 1, "plt.close() # HACK")
```
Comment (androj07) [inline date_a_scientist/__init__.py:55]: Is it always necessary or sometimes if the plt.close is missing?

Comment (androj07) [inline date_a_scientist/__init__.py:55]: Is it always necessary or sometimes if the plt.close is missing?

File: date_a_scientist/__init__.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        if isinstance(result, str) and "exports/charts" in result:
+            for row in result.split("\n"):
+                row = row.strip()
+                match = re.search(pattern, row)
+                if match:
+                    path = match.group()
+
+                    try:
+                        from IPython.display import \
```
Comment (androj07) [inline date_a_scientist/__init__.py:100]: Could the import be defined at the beginning of the file ?

File: date_a_scientist/__init__.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        else:
+            return result
+
+    def code(self, q: str) -> Any:
+        code = self._agent.generate_code(self._query(q))
+        code = code.replace("dfs[0]", "df")
+
+        try:
+            from IPython.display import HTML  # type: ignore[import]
```
Comment (androj07) [inline date_a_scientist/__init__.py:117]: Could the import be defined at the beginning of the file ?

File: date_a_scientist/__init__.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            return HTML(f'<style>{formatter.get_style_defs(".highlight")}</style>{highlighted_code}')
+
+        except ImportError:
+            return code
+
+    @cached_property
+    def _agent(self):
+        self._assure_llm_openai_api_token()
+
+        llm = _CustomOpenAI(model="gpt-4o", api_token=self._llm_openai_api_token)
```
Comment (androj07) [inline date_a_scientist/__init__.py]: I guess the model should take the `self._llm_openai_model`

**Review by androj07 — APPROVED** [review]
I've left a few comments.

Another one:
I've noticed that some users do not have access to the gpt-4o model and they may receive and error on attempt to use the model

**Comment by maciejATcosphere** [comment]
> I've left a few comments.
> 
> Another one: I've noticed that some users do not have access to the gpt-4o model and they may receive and error on attempt to use the model

You're super right: https://platform.openai.com/docs/guides/rate-limits?context=tier-free

File: date_a_scientist/__init__.py (hunk 3 above)
Comment (maciejATcosphere) [reply date_a_scientist/__init__.py:55]: basicaly when it's missing but adding extra doesn't harm

File: date_a_scientist/__init__.py (hunk 4 above)
Comment (maciejATcosphere) [reply date_a_scientist/__init__.py:100]: nope, this is a trick. If we want this class to work in notebooks and normal python scripts then different outputs should be returned. We detect if we're under jupyter lab / notebook with the import

File: date_a_scientist/__init__.py (hunk 5 above)
Comment (maciejATcosphere) [reply date_a_scientist/__init__.py:117]: see above

File: date_a_scientist/__init__.py (hunk 6 above)
Comment (maciejATcosphere) [reply date_a_scientist/__init__.py]: of course :-P

**Pull request merged**